url27a7ae3508_serkan_ozal_ocean_of_memories_tgz-pJ8-com_zeroturnaround_rebellabs_oceanofmemories_article1_objectlayout_ObjectMemoryLayoutDemoJ8/src/com/zeroturnaround/rebellabs/oceanofmemories/common/util/ClasspathUtil.java

```diff
@@ -0,0 +1,207 @@
+        }
+        fullClasspath = classpathBuilder.toString().replace("%20", " ");
+    }
+
+    private static Set<URL> findClasspathUrls() {
+        Set<URL> urls = new HashSet<URL>();
+
+        try {
+            String[] classpathProperties = System
+                    .getProperty("java.class.path").split(File.pathSeparator);
+            for (String classpathProperty : classpathProperties) {
+                urls.add(new File(classpathProperty).toURI().toURL());
+            }
+        } catch (MalformedURLException e) {
+            logger.error(
+                    "Error occured while getting classpath from system property \"java.class.path\"",
+                    e);
+        }
+
+        String surefireProperty = System
+                .getProperty("surefire.test.class.path");
+        if (surefireProperty != null && surefireProperty.length() > 0) {
+            try {
+                String[] surefireClasspathProperties = surefireProperty
+                        .split(File.pathSeparator);
+                for (String surefireClasspathProperty : surefireClasspathProperties) {
+                    urls.add(new File(surefireClasspathProperty).toURI()
+                            .toURL());
+                }
+            } catch (MalformedURLException e) {
+                logger.error(
+                        "Error occured while getting classpath from system property \"surefire.test.class.path\"",
+                        e);
+            }
+        }
+
+        // Start with Current Thread's loader
+        ClassLoader ctxLoader = Thread.currentThread().getContextClassLoader();
+        ClassLoader loader = ctxLoader;
+        while (loader != null) {
+            urls.addAll(findClasspathsByLoader(loader));
+            loader = loader.getParent();
+        }
+
+        // Also start with this classes's loader, in some environment this can
+        // be different than the current thread's one
+        ClassLoader appLoader = ClasspathUtil.class.getClassLoader();
+        loader = appLoader;
+        while (loader != null) {
+            urls.addAll(findClasspathsByLoader(loader));
+            loader = loader.getParent();
+        }
+
+        ClassLoader sysLoader = ClassLoader.getSystemClassLoader();
+        loader = sysLoader;
+        while (loader != null) {
+            urls.addAll(findClasspathsByLoader(loader));
+            loader = loader.getParent();
+        }
+
+        Map<URL, URL> replaceURLs = new HashMap<URL, URL>();
+        Set<URL> derivedUrls = new HashSet<URL>();
+        for (URL url : urls) {
+            if (url.getProtocol().startsWith("vfs")) {
+                try {
+                    URLConnection conn = url.openConnection();
+                    Object virtualFile = conn.getContent();
+                    if (virtualFile.getClass().getName()
+                            .equals("org.jboss.vfs.VirtualFile")) {
+                        File file = (File) virtualFile.getClass()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)